### PR TITLE
Connect react-grid-layout changes with dashboard store

### PR DESCRIPTION
**Pull request by LukeTillman**
For each panel group, this PR splits the item state inside that panel group into two separate pieces of state:

- `itemLayouts`: an array of the layout information (`x`, `y`, `w`, `h`, and a unique ID `i`) which corresponds 1:1 with the model that `react-grid-layout` expects
- `itemPanelKeys`: a map of the unique ID of the item in `itemLayouts` -> the panel key referenced by that panel

By splitting this state up, I was then able to connect `react-grid-layout` with the store in a pretty straightforward manner. Its `onLayoutChange` handler can now just overwrite the `itemLayouts` property on a panel group every time a move/resize happens, while the content reference (i.e. panel key) remains stable. This also means that drag and drop actions should now be reflected in the object that's output for conversion to JSON. 🥳

### CODE REVIEW

File: ui/dashboards/src/context/DashboardProvider/dashboard-provider-api.ts [hunk 1]
```diff
@@ -136,12 +137,11 @@ export function useDeletePanelGroupDialog() {
...
  */
 export function usePanel(panelGroupItemId: PanelGroupItemId) {
-  const { panelGroupId, itemIndex } = panelGroupItemId;
+  const { panelGroupId, panelGroupLayoutId } = panelGroupItemId;
 
   const panel = useDashboardStore((store) => {
-    const panelRef = store.panelGroups[panelGroupId]?.items[itemIndex]?.content;
-    if (panelRef === undefined) return;
-    const panelKey = getPanelKeyFromRef(panelRef);
+    const panelKey = store.panelGroups[panelGroupId]?.itemPanelKeys[panelGroupLayoutId];
```
Comment (sjcobb) [inline ui/dashboards/src/context/DashboardProvider/dashboard-provider-api.ts:143]: Oh this reminds me, not sure if it's still an issue, but last week I saw some weird behavior where if I gave two panels the same name, there would be a collision and subsequent changes to one would effect the other (I think I created a StatChart  and GaugeChart, then they both became GaugeCharts with duplicate names / descriptions).

**Review by sjcobb — APPROVED** [review]
🚀

File: ui/dashboards/src/context/DashboardProvider/dashboard-provider-api.ts (hunk 1 above)
Comment (LukeTillman) [reply ui/dashboards/src/context/DashboardProvider/dashboard-provider-api.ts:143]: Yeah, that's been in the back of my mind as something we need to fix for panel keys. Definitely still an issue.

File: ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts [hunk 2]
```diff
@@ -57,17 +63,24 @@ export type PanelGroupId = number;
...
   id: PanelGroupId;
-  items: GridItemDefinition[];
   isCollapsed: boolean;
   title?: string;
+  itemLayouts: PanelGroupLayout[];
+  itemPanelKeys: Record<PanelGroupLayoutId, string>;
 }
 
+export interface PanelGroupLayout extends Layout {
+  i: PanelGroupLayoutId;
```
Comment (eunicorn) [inline ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts:73]: can we call this `id`?

File: ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts [hunk 3]
```diff
@@ -57,17 +63,24 @@ export type PanelGroupId = number;
...
 export interface PanelGroupDefinition {
   id: PanelGroupId;
-  items: GridItemDefinition[];
   isCollapsed: boolean;
   title?: string;
+  itemLayouts: PanelGroupLayout[];
+  itemPanelKeys: Record<PanelGroupLayoutId, string>;
 }
 
+export interface PanelGroupLayout extends Layout {
```
Comment (eunicorn) [inline ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts:72]: can we rename this to PanelGroupItemLayout? PanelGroupLayout makes it seems like its the layout for the whole panel group itself

Comment (eunicorn) [reply ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts:72]: naming is hard 😆

File: ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts [hunk 4]
```diff
@@ -76,18 +89,36 @@ export interface PanelGroupItemId {
...
-    const id = generateId();
-    panelGroups[id] = {
-      id,
-      items: layout.spec.items,
+    const itemLayouts: PanelGroupDefinition['itemLayouts'] = [];
+    const itemPanelKeys: PanelGroupDefinition['itemPanelKeys'] = {};
+
+    // Split layout information from panel keys to make it easier to update just layouts on move/resize of panels
+    for (const item of layout.spec.items) {
+      const panelGroupLayoutId = generateId().toString();
```
Comment (eunicorn) [inline ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts:101]: i am thinking if we could somehow generate/enforce unique `panelRef`, then we wouldn't have to keep track of unique id and panelRef

File: ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts (hunk 2 above)
Comment (LukeTillman) [reply ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts:73]: This comes from `react-grid-layout` (I basically just redefined the type from `string` so it was more obvious what it actually is), so unfortunately has to stay `i` unless we want to have to transform back and forth with their format.

File: ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts (hunk 3 above)
Comment (LukeTillman) [reply ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts:72]: Yup, good call.

File: ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts (hunk 4 above)
Comment (LukeTillman) [reply ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts:101]: Yeah, we're still dealing with the whole "panels can be referenced multiple times" thing, so a `panelRef` isn't guaranteed to be unique. 🙃

**Review by eunicorn — APPROVED** [review]

**Pull request merged**